### PR TITLE
[FLINK-4927] [yarn]Implement FLI-6 YARN Resource Manager

### PR DESCRIPTION
**Pull request by shuai-xu**
This pr is for issue #[4927](https://issues.apache.org/jira/browse/FLINK-4927). It is based on #2878 , please refer to #2878 for YarnTaskExecutorRunner.

The Flink YARN Resource Manager communicates with YARN's Resource Manager to acquire and release containers. And it will start the allocated containers.

### CODE REVIEW

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+			}
+		}
+		if (numPendingContainerRequests <= 0) {
+			resourceManagerClient.setHeartbeatInterval(yarnHeartbeatIntervalMillis);
+		}
+	}
+
+	@Override
+	public void onShutdownRequest() {
+		// Nothing to do
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: When do we get this request?

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+	}
+
+	@Override
+	public void onNodesUpdated(List<NodeReport> list) {
+		// We are not interested in node updates
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: What kind of information is sent on a nodes update message?

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+			resourceManagerClient.registerApplicationMaster(hostPort.f0, hostPort.f1, getAddress());
+		} catch (Exception e) {
+			LOG.info("registerApplicationMaster fail", e);
+		}
+
+		// create the client to communicate with the node managers
+		nodeManagerClient = NMClient.createNMClient();
+		nodeManagerClient.init(yarnConfig);
+		nodeManagerClient.start();
+		nodeManagerClient.cleanupRunningContainersOnStop(true);
```
Comment (shuai-xu) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: There are some problems if we do so. First the AMRMClientAsync need a CallbackHandler, but Dispatcher and YarnResourceManager need different CallbackHandlers. And I think what we need in Dispatcher to allocate containers for new JobMasters should be YarnClient.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java (hunk 2 above)
Comment (shuai-xu) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: The yarn master will send nodes who's status is changed by this call. We can get enough by the notices of containers, so need not process this call.

**Comment by StephanEwen** [comment]
I have started to and merge this together with #2744 (both are needed together to compile).
Will file a few JIRA issues with followups.

**Comment by StephanEwen** [comment]
Manually merged to the FLIP-6 branch.

**Pull request closed without merging**